Svc/DpCatalog/DpCatalog.cpp

```diff
@@ -102,10 +105,235 @@ namespace Svc {
         this->m_initialized = true;
     }
 
+    void DpCatalog::resetBinaryTree() {
+        // initialize data structures in the free list
+        this->m_freeListHead = static_cast<DpBtreeNode*>(this->m_memPtr);
+        for (FwSizeType slot = 0; slot < this->m_numDpSlots; slot++) {
+            // overlay new instance of the DpState entry on the memory
+            (void) new(&this->m_freeListHead[slot]) DpBtreeNode();
+            this->m_freeListHead[slot].left = nullptr;
+            this->m_freeListHead[slot].right = nullptr;
+            // link the free list
+            if (slot > 0) {
+                this->m_freeListHead[slot-1].left = &this->m_freeListHead[slot];
+            }
+        }
+        // set the foot of the free list
+        this->m_freeListFoot = &this->m_freeListHead[this->m_numDpSlots - 1];
+        // clear binary tree
+        this->m_dpTree = nullptr;
+        // reset number of records
+        this->m_numDpRecords = 0;
+    }
+
+    void DpCatalog::resetStateFileData() {
+        // clear state file data
+        for (FwSizeType slot = 0; slot < this->m_numDpSlots; slot++) {
+            this->m_stateFileData[slot].used = false; 
+            this->m_stateFileData[slot].visited = false; 
+            (void) new(&this->m_stateFileData[slot].entry.record) DpRecord();
```

Comment:
Is it a good idea to placement-new into an existing entry?  It seems the construction of the entry should imply the construction of the record.